### PR TITLE
Add functional-match-pattern support to match statement

### LINKED ISSUE #24293: Add functional-match-pattern to match statement

### PR DESCRIPTION
**Pull request by rdulmina**
## Purpose
$Subject

Fixes #24293 and #24388

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 1]
```diff
@@ -12744,6 +12754,251 @@ private STNode parseFieldMatchPatternRhs(SyntaxKind nextTokenKind) {
...
+     *     <br/>
+     *     type-reference := identifier | qualified-identifier
+     *     <br/>
+     *     arg-list-match-pattern := positional-arg-match-patterns [, other-arg-match-patterns]
+     *    | other-arg-match-patterns
+     *</p>
+     *
+     * @return Parsed functional match pattern node.
+     */
+    private STNode parseFunctionalMatchPattern(STNode typeRef) {
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:12814]: I think we can simplify this method and `parseOtherArgMatchPatterns` method, if we follow the same approach as `parseArgList` method.

A further simplified/cleanup version can be found in [parseArgListBindingPatterns](https://github.com/SupunS/ballerina/blob/7a89afd76da3534524b2caa582d246ded08305f1/compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java#L13626) (still WIP). I would suggest let's do similar to this one.

File: compiler/ballerina-treegen/src/main/resources/syntax_tree_descriptor.json [hunk 2]
```diff
@@ -3938,6 +3938,70 @@
...
+                },
+                {
+                    "name": "positionalArgMatchPatternsNode",
+                    "type": "Node",
+                    "occurrences": "MULTIPLE_SEPARATED"
+                },
+                {
+                    "name": "otherArgMatchPatternsNode",
+                    "type": "OtherArgMatchPatternsNode"
+                },
```
Comment (SupunS) [inline compiler/ballerina-treegen/src/main/resources/syntax_tree_descriptor.json:3962]: The comma in between the two lists are not captured? 
Would it be better to merge these two lists in to one (similar to arguments list in a function call)?

Comment (rdulmina) [reply compiler/ballerina-treegen/src/main/resources/syntax_tree_descriptor.json:3962]: That comma is captured to the end of positionalArgMatchPatterns list.
Will look into that

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java (hunk 1 above)
Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:12814]: sure will look into that :-)

Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:12814]: Improved

File: compiler/ballerina-treegen/src/main/resources/syntax_tree_descriptor.json (hunk 2 above)
Comment (rdulmina) [reply compiler/ballerina-treegen/src/main/resources/syntax_tree_descriptor.json]: Changed

**Pull request merged**